src/Middleware/CacheMiddleware.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace Dumbo\Middleware;
+
+use Dumbo\Context;
+use Psr\Http\Message\ResponseInterface;
+
+class CacheMiddleware
+{
+    const HTTP_NOT_MODIFIED = 304;
+
+    public static function withHeaders(
+        string $type = 'private',
+        bool   $mustRevalidate = false,
+        int    $maxAge = 86400,
+        bool   $strictEtag = false,
+    ): callable
+    {
+        return function (Context $ctx, callable $next) use ($type, $mustRevalidate, $maxAge, $strictEtag): ResponseInterface {
+
+            $request = $ctx->req;
+
+            if ($request->method() != 'GET') {
+                return $next($ctx);
+            }
+
+            $etag = $request->header('If-None-Match') ?: sprintf('"W/%s"', md5($ctx->req->getUri()));
```

Comment:
Might it read better if we created a function for generating the etag?

```php
$etag = $ctx->req->header('If-None-Match') ?: self::generateEtag($ctx, $strictEtag);
```

Then have something like this inside the middleware (which means we can remove the changes to RequestWrapper 🥳 

```php
private static function generateEtag(Context $ctx, bool $strict): string
    {
        $identifier = $strict
            ? $ctx->req->method() . $ctx->req->path() . serialize($ctx->req->query())
            : $ctx->req->path();

        return sprintf('W/"%s"', md5($identifier));
    }
```

Reply:
Got it, thanks for the feedback !
 
I start when I have a bit of time🫡